tools/gristool.8
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.14
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "GRISTOOL 8"
.TH GRISTOOL 8 "2005-03-03" "0.1-DEVEL" "Grist Grelist Policy Server Package"
.SH "NAME"
\&\fBgristool\fR \- grist greylist database managment tool
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
gristool.pl [\fB\s-1OPTIONS\s0\fR] [\fB\s-1COMMAND\s0\fR] \fBarguments\fR...
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBGristool\fR is part of the grist greylist policy server toolset. \fBGristool\fR provides a simplified 
interface to the greylist database and is required for common database maintaince procedures.
.PP
This man page was written very late one night, or early one morning depedning how you want to look
at it. A rewrite and restructure is pending but the beef of the information is acurate, and if I 
goofed the utility will let you know. =)
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-c\fR, \fB\-\-conf\fR" 6
.IX Item "-c, --conf"
Specify the full path and name of the grist configuration file. 
.Sp
Default is: /usr/local/etc/grist.conf
.IP "\fB\-p\fR, \fB\-\-pretend\fR" 6
.IX Item "-p, --pretend"
Will prevent any modifications from beign made to the database. Only applies to operations which would 
alter the database other commands will simply ignore it.
.IP "\fB\-v\fR, \fB\-\-verbose\fR" 6
.IX Item "-v, --verbose"
Will cause \fBgristool\fR to print various messages to the screen throughout the course of execution. This
option is usually only useful for troubleshooting.
.IP "\fB\-V\fR, \fB\-\-version\fR" 6
.IX Item "-V, --version"
Display version information and exit.
.SH "COMMANDS"
.IX Header "COMMANDS"
\&\fBGristool\fR has serveral modes of operation which are accessable through command subsets. Below is a list
of commands that are currently implemented and a detailed look at thier usage.
.Sh "prune \s-1AGE\s0 <dead\-requests>"
.IX Subsection "prune AGE <dead-requests>"
.RS 6
The \fBprune\fR command is used to remove request records from the greylist database. 
.Sp
All requests older than or equal to \fB\s-1AGE\s0\fR will be removed from the database. \fB\s-1AGE\s0\fR may be specified as 
seconds, minutes, hours, or days by appending 's', 'm', 'h' or 'd' as a value hint. 
.Sp
Examples of valid \fB\s-1AGE\s0\fR values:
.Sp
.Vb 4
\&        45d     translates to 45 days or <xxx> seconds
\&        3h      translates to 3 hours or <xxx> seconds
\&        45m     translates to 45 minutes or <xxx> seconds
\&        100s    translates to 100 seconds
.Ve
.RE
.RS 6
.Sp
An \fB\s-1AGE\s0\fR value without a trailing 's', 'm', 'h' or 'd' is considered invalid.
.Sp
If \fBdead-requests\fR is specified only records which have a seen count of zero are removed from the
database. 
.RE
.Sh "check \s-1FIELD\s0 \s-1VALUE\s0"
.IX Subsection "check FIELD VALUE"
.RS 6
The \fBcheck\fR command allows you to query the greylist database on one field for an exact value match
and display the results to the console. 
.Sp
It is planned to extend the functionality of this command to allow for similar matching and querying
on multiple fields. In addition other commands for more complex analysis queries are on the drawing
board. 
.Sp
Valid \fB\s-1FIELD\s0\fR values are:
.Sp
.Vb 4
\&        from|sender     
\&        to|recipient
\&        ip|sourceip
\&        host|hostname
.Ve
.Sp
Examples:
.Sp
.RS 6
To seach for all requests sent from joe@mydomain.com:
.Sp
.RS 6
\&\fBgristool check from joe@mydomain.com\fR
.RE
.RE
.RS 6
.Sp
To search for all requests for maria@herdomain.com:
.Sp
.RS 6
\&\fBgristool check to maria@herdomain.com\fR
.RE
.RE
.RS 6
.RE
.RE
.RS 6
.Sp
The output of this command is very self explainatory i will discuss it later.. maybe.
.SH "GREYLIST DATABASE MAINTAINCE"
.IX Header "GREYLIST DATABASE MAINTAINCE"
When implementing a greylisting solution for a high traffic mail server the greylist database will grow very quickly. 
Even for low traffic mail servers the database will require regular pruning though hourly might be excessive.
It is recommended that you schedule \fBgristool\fR to run at regular intervals using a schedualing daemon such as
\&\fBcron\fR to remove any requests which are considered 'dead'.
.Sp
An example \fBcron\fR entry to run \fBgristool\fR every hour and remove all dead request records older than 5 
hours would be as follows:
.Sp
.RS 6
0 * * * * /usr/local/bin/gristool prune 5h dead-requests
.RE
.RE
.RS 6
.Sp
It is also recommended that you schedule a monthly run to remove all records over a specific age. 
.Sp
An example \fBcron\fR entry to run \fBgristool\fR once a month and remove all request records older than 30 days would 
be as follows:
.Sp
.RS 6
* * 1 * * /usr/local/bin/gristool prune 30d
.RE
.RE
.RS 6
.SH "AUTHORS"
.IX Header "AUTHORS"
\&\s-1TODO\s0
.SH "REPORTING BUGS"
.IX Header "REPORTING BUGS"
\&\s-1TODO\s0
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
\&\s-1TODO\s0
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\s-1TODO\s0


ChangeLog
# do not edit -- automatically generated by arch changelog
# arch-tag: automatic-ChangeLog--vcs@digital-fallout.com--2005/grist--devel--0.1
#

2005-03-26 04:30:29 GMT	Michael Hubbard <mhubbard@digital-fallout.com>	patch-18

    Summary:
      added --conf option to allow for user definable configuration path
    Revision:
      grist--devel--0.1--patch-18

    added --conf option to allow for user definable configuration path

    modified files:
     src/main.c


2005-03-26 03:56:49 GMT	Michael Hubbard <mhubbard@digital-fallout.com>	patch-17

    Summary:
      added postgresql support
    Revision:
      grist--devel--0.1--patch-17

    added postgresql support
    tweaked the instert query to be cross-database independant or 'standard'
    

    modified files:
     src/db_sql.c


2005-03-26 02:47:50 GMT	Michael Hubbard <mhubbard@digital-fallout.com>	patch-16

    Summary:
      added support for mysql servers
    Revision:
      grist--devel--0.1--patch-16

    added support for mysql servers

    modified files:
     src/db_sql.c src/main.c


2005-03-26 01:11:50 GMT	Michael Hubbard <mhubbard@digital-fallout.com>	patch-15

    Summary:
      updated libdbi help
    Revision:
      grist--devel--0.1--patch-15

    update libdbi help string
    
    
    

    modified files:
     configure.ac


2005-03-09 01:17:14 GMT	Michael Hubbard <mhubbard@digital-fallout.com>	patch-14

    Summary:
      tweaks to sql query loops
    Revision:
      grist--devel--0.1--patch-14

    
    

    modified files:
     Makefile.am src/db_sql.c


2005-03-08 23:57:11 GMT	Michael Hubbard <mhubbard@digital-fallout.com>	patch-13

    Summary:
      corrected dynamic version reading
    Revision:
      grist--devel--0.1--patch-13

    VERSION can now also contain a -devel suffix.

    modified files:
     Makefile.am autogen.sh configure.ac src/Makefile.am src/main.c


2005-03-08 23:15:05 GMT	Michael Hubbard <mhubbard@digital-fallout.com>	patch-12

    Summary:
      added support for new configuration options
    Revision:
      grist--devel--0.1--patch-12

    added support for 'rq_cooldown' and 'rq_defer_msg'

    modified files:
     gristool/Grist/Config.pm


2005-03-08 23:11:42 GMT	Michael Hubbard <mhubbard@digital-fallout.com>	patch-11

    Summary:
      corrected syslog message output
    Revision:
      grist--devel--0.1--patch-11

    log messages now go to the correct logging levels. i guess LOG_*|LOG_* will not send to
    multiple levels as expected.

    modified files:
     src/main.c


2005-03-08 23:00:50 GMT	Michael Hubbard <mhubbard@digital-fallout.com>	patch-10

    Summary:
      possible solution to sqlite query errors
    Revision:
      grist--devel--0.1--patch-10

    sqlite seems to tweak under load, added a 'query attempt' loop to retry a max of 10 times
    before reporting a real failure.

    modified files:
     src/db_sql.c


2005-03-08 22:35:39 GMT	Michael Hubbard <mhubbard@digital-fallout.com>	patch-9

    Summary:
      build requirements tweaking
    Revision:
      grist--devel--0.1--patch-9

    improved check for libdbi
    added check for 'stdarg.h' header

    modified files:
     configure.ac


2005-03-08 22:14:27 GMT	Michael Hubbard <mhubbard@digital-fallout.com>	patch-8

    Summary:
      script now updates VERSION
    Revision:
      grist--devel--0.1--patch-8

    VERSION file is updated before building of configure, in the future would be nice
    to have a tla-hook do this on commits

    modified files:
     autogen.sh


2005-03-08 22:12:39 GMT	Michael Hubbard <mhubbard@digital-fallout.com>	patch-7

    Summary:
      version now read from VERSION file when configure is built.
    Revision:
      grist--devel--0.1--patch-7

    application version is now read from the VERSION file by configure.ac

    modified files:
     configure.ac


2005-03-07 02:53:30 GMT	Michael Hubbard <mhubbard@digital-fallout.com>	patch-6

    Summary:
      command line operation tweaks
    Revision:
      grist--devel--0.1--patch-6

    implemented '--version' to display grist version
    added a nicer usage message if an unknown argument is passed

    modified files:
     Makefile.am configure.ac src/Makefile.am src/main.c


2005-03-07 02:35:26 GMT	Michael Hubbard <mhubbard@digital-fallout.com>	patch-5

    Summary:
      add 'gristool' to project tree
    Revision:
      grist--devel--0.1--patch-5

    gristool is a perl script and a set of perl modules for managing the greylist database

    new files:
     gristool/.arch-ids/=id gristool/.arch-ids/gristool.pl.id
     gristool/Grist/.arch-ids/=id
     gristool/Grist/.arch-ids/Check.pm.id
     gristool/Grist/.arch-ids/Config.pm.id
     gristool/Grist/.arch-ids/Prune.pm.id gristool/Grist/Check.pm
     gristool/Grist/Config.pm gristool/Grist/Prune.pm
     gristool/docs/.arch-ids/=id
     gristool/docs/.arch-ids/gristool.pod.id
     gristool/docs/gristool.pod gristool/gristool.pl

    new directories:
     gristool gristool/.arch-ids gristool/Grist
     gristool/Grist/.arch-ids gristool/docs gristool/docs/.arch-ids


2005-03-07 02:33:53 GMT	Michael Hubbard <mhubbard@digital-fallout.com>	patch-4

    Summary:
      complete rewrite
    Revision:
      grist--devel--0.1--patch-4

    complete rewrite, cleaner and safer implementation.

    new files:
     src/.arch-ids/config.c.id src/.arch-ids/db_sql.c.id
     src/.arch-ids/db_sql.h.id src/config.c src/db_sql.c
     src/db_sql.h

    modified files:
     Makefile.am src/Makefile.am src/grist.h src/main.c


2005-02-26 05:16:57 GMT	Michael Hubbard <mhubbard@digital-fallout.com>	patch-3

    Summary:
      added '--enable-debug' support for configure
    Revision:
      grist--devel--0.1--patch-3

    debug code can now be turned on by passing '--enable-debug' to the configure script
    added -DMEMWATCH to CFLAGS when '--enable-memwatch' is passed to configure

    modified files:
     configure.ac src/Makefile.am



gristool/gristool.pl
#!/usr/bin/perl -w
#
# file: gristool.pl 
# grist - greylist policy server database managment tool
#
# Copyright (C) 2005 Michael Hubbard <mhubbard@digital-fallout.com> 
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#
# Authors: Michael Hubbard <mhubbard@digital-fallout.com>
#

use strict;
use Getopt::Long qw(:config gnu_getopt );

use Grist::Config;
use Grist::Prune;
use Grist::Check;

my $appname = `basename $0`; chomp($appname);
my $appver  = '0.1-DEVEL';
my $helpstr = "Try '$appname --help' for more information.";

my $grist_conf = '../src/grist.conf';
my $config = new Config();

my %opts = ( 'verbose' => 0, 'pretend' => 0, 'conf' => $grist_conf );
my @args = ( );

# trap warn signal
$SIG{__WARN__} = sub {
	if ( $_[0] =~ /^Unknown option:|^Option/ ) {
		# getopt option error, make it pretty and use out usage()
		my $message = lc($_[0]); chomp($message);
		usage($helpstr, "$appname: $message");
		return
	} else {
		# handle warning as usual
		warn @_;
	}
};

# version - displays application version and exits
#
sub version {
	print "$appname v$appver\n";
}

# usage - displays command line usage help
#
sub usage {
	my ( $message ) = shift;
	my ( $badopt  ) = shift;
	my ( $cmd     ) = shift;

	if ( $badopt ) { print "$badopt\n"; }
	
	if ( !$cmd ) { 
		$cmd = ''; 
		print "Usage: $appname [OPTIONS] [COMMAND] arguments...\n";
	};

	if ( $cmd eq 'help' ) {
		print "Usage: $appname [OPTIONS] [COMMAND] arguments...\n";
		print "Grist greylist database management tool.\n";
		print "\n";
		print "Options:\n";
		print "  -c, --conf    \tfull path and name of the grist configuration file.\n";
		print "  -p, --pretend \tdisplay what would be done, will not alter the database.\n";
		print "  -v, --verbose \tbe verbose.\n";
		print "  -V, --version \tdisplay version information and exit.\n";
		print "\n";
		print "Commands:\n";
		print "  prune AGE <dead-requests>\tremove requests older than AGE.\n";
		print "  check FIELD VALUE        \tquery FIELD for VALUE and display matches.\n";
		print "\n";
	} elsif ( $cmd eq 'prune' ) {
		print "Usage: $appname [OPTIONS] prune AGE <dead-requests>\n";
		print "$helpstr\n";
	} elsif ( $cmd eq 'check' ) {
		print "Usage: $appname [OPTIONS] check FIELD VALUE\n";
		print "$helpstr\n";
	}
	
	if ( $cmd ne '' ) {
		print "\n";
		print "For more detailed information try `man gristool`.\n";
		print "\n";
		print "Report bugs to <bugs\@digital-fallout.com>.\n";
	}		
	
	if ( $message ) { print "$message\n"; }

	exit(0);
}

# save option arguments
#
sub storeArgument {
	$args[++$#args] = shift;
}

#
# end subs, application code begins
#

# no arguments, show usage.
if ( $#ARGV == -1 ) { usage( $helpstr ); }
	
GetOptions(\%opts,
	   'conf|c=s',
	   'verbose|v',
	   'pretend|p',
	   'V|version', => sub { version(); },
	   'help|h|?' => sub { usage('','','help'); },
	   '<>' => \&storeArgument
	  );

if ( $opts{'verbose'} ) {
	print "$appname: welcome, guess all went well so far.\n";
	print "$appname: obviously running in verbose mode.\n";
	if ( $opts{'pretend'} ) { print "$appname: running in pretend mode.\n"; }
	print "$appname: using grist configuration: $opts{'conf'}\n";
}

# final sanity check on our wanted command in $args[0]
if ( !$args[0] ) { usage($helpstr); } 

# check our $opts{'conf'}
if ( ! -f $opts{'conf'} ) { print STDERR "error: configuration file not found '$opts{'conf'}'.\n"; exit(1); }

$config->setFilename($opts{'conf'});
if ( !$config->parse() ) {
	exit(1);
}

if ( $opts{'verbose'} ) { print "$appname: checking for existance of the sqlite database.\n"; }
if ( $config->get('db_driver') eq 'SQLite' && (! -f $config->get('db_name')) ) {
	print STDERR "error: greylist database '".$config->get('db_name')."' does not exist.\n";
	exit(0);
}

# build database connection string
my $dsn = sprintf('DBI:%s:dbname=%s;host=%s;port=%s',
			$config->get('db_driver'),
			$config->get('db_name'),
			$config->get('db_host'),
			$config->get('db_port')
		  );

if ( $opts{'verbose'} ) { print "$appname: using $dsn\n"; }

# check for command
if ( $opts{'verbose'} ) {
	print "$appname: command arguments follow.\n";
	for( my $idx=0; $idx<=$#args; $idx++ ) {
		print "$appname: \$args[$idx] = $args[$idx]\n";
	}
}

my $command = shift(@args);
if ( lc($command) eq 'prune' ) {
	
	if ( $opts{'verbose'} ) { print "$appname: command 'prune' requested, handing over control ...\n"; }
	# check for prune
	my $prune = new Prune( $opts{'verbose'}, $opts{'pretend'} );
	if ( !$prune->checkArguments(@args) )  { usage('','','prune'); }

	if (! $prune->perform( $dsn, $config->get('db_username'), $config->get('db_password') ) ) {
		if ( $opts{'verbose'} ) { print "$appname: exiting with error condition.\n"; }
		exit(1);
	} 
	
} elsif ( lc($command) eq 'check' ) {
	
	if ( $opts{'verbose'} ) { print "$appname: command 'check' requested, handing over control ...\n"; }
	my $check = new Check( $opts{'verbose'}, $opts{'pretend'} );
	if ( !$check->checkArguments(@args) ) { usage('','','check'); }

	if ( !$check->perform($dsn, $config->get('db_username'), $config->get('db_password')) ) {
		if ( $opts{'verbose'} ) { print "$appname: exiting with error condition.\n"; }
		exit(1);
	} 

} else {
	usage($helpstr, "error: unknown command requested '$args[0]'");
}

if ( $opts{'verbose'} ) { print "$appname: exiting, success.\n"; }

exit(0);
